File: Lecture_Code/Week3/BOJ_10816.py
# BOJ_10816.py
import sys

input = sys.stdin.readline

# lower bound
def BinarySearch2(arr: list[int], target: int) -> int:
	left, right = -1, len(arr)
	while (left + 1 < right):
		mid = (left + right) // 2
		if (arr[mid] < target):
			left = mid
		else:
			right = mid

	return right

# upper bound
def BinarySearch3(arr: list[int], target: int) -> int:
	left, right = -1, len(arr)
	while (left + 1 < right):
		mid = (left + right) // 2
		if (arr[mid] <= target):
			left = mid
		else:
			right = mid

	return right

n = int(input())
arr = list(map(int, input().split()))
arr.sort()

m = int(input())
targets = list(map(int, input().split()))
for target in targets:
	ans = BinarySearch3(arr, target) - BinarySearch2(arr, target)
	print(ans, end = ' ')
print()

File: Binary_Search/BOJ_10816.cpp
// BOJ_10816.cpp
#include <iostream>

#include <algorithm>

using namespace std;

// lower_bound
int BinarySearch2(int arr[], int size, int target) {
	int left = -1, right = size;
	while (left + 1 < right) {
		int mid = (left + right) / 2;
		if (arr[mid] < target)
			left = mid;
		else
			right = mid;
	}

	return right;
}

// upper_bound
int BinarySearch3(int arr[], int size, int target) {
	int left = -1, right = size;
	while (left + 1 < right) {
		int mid = (left + right) / 2;
		if (arr[mid] <= target) // 같을 땐 범위를 오른쪽 범위로
			left = mid;
		else
			right = mid;
	}

	return right;
}

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	int n, arr[500'000];
	cin >> n;

	for (int i = 0; i < n; i++)
		cin >> arr[i];

	sort(arr, arr + n);

	int m;
	cin >> m;
	for (int i = 0; i < m; i++) {
		int target;
		cin >> target;

		int ans = BinarySearch3(arr, n, target) - BinarySearch2(arr, n, target);
		cout << ans << ' ';
	}
	cout << '\n';

	return 0;
}